### DISCUSSION TITLE
using of functions with mpdf

### DISCUSSION THREAD

**Discussion by sscherger** [mpdf/mpdf#2020] [General]
Hi,

so far I used TCPDF for invoices, Now we change to MPDF to create e-invoices. So far it was possible to print a normal pdf sample, and we also have access to our MySQL data. That looks fine.
But now I want to implement a function for example to have a field for the addresses on the top of the page. 
How do I declare a function in MPDF correctly? There seems to be differences to TCPDF, but I can't find the difference. Still tried a lot...

Another question is: Do we have to use the HTML, or can we also use the commands from TCPDF, like $pdf->MultiCell(0,5,'"'.($result->Whatever).'"',0,'L',0); ?

Thank you for your support!!

Regards
Stefan

**Comment by finwe** [comment]
Care to consult the documentation? https://mpdf.github.io/reference/mpdf-functions/multicell.html

**Comment by finwe** [comment]
As for the first question, I am afraid I don't understand what are you trying to accomplish.

**Comment by sscherger** [comment]
Shame on me.. I did not have a closer look into the manual regarding that functions, because so far I was not able to implement a self written function. That is what I meant with the first question.  A construction like:

```
class print extends \Mpdf\Mpdf{
	function Address($x,$y){		
		$this->SetY($y); 	
		$this->SetX($x);
		$this->MultiCell(85,3, ($company),0,'L',0,1); ...
}
$mpdf->address($x,$y)  // this is my function
```

In the meantime I have such a function that does not cause an error, but when I open the pdf then it's a blank paper. Thought I used it like before with TCPDF. Or do I have to initialize the MPDF in another way like a TCPDF? 

Thanks for help!!

**Reply by finwe** [reply]
Why would you extend mPDF when you can much more cleanly just encapsulate the object into a reusable helper function:

```
function printAddress(Mpdf $mpdf, $x, $y, $company) 
{
    $mpdf->SetY($y); 	
    $mpdf->SetX($x);
    $mpdf->MultiCell(85,3, ($company),0,'L',0,1);
}
```

You should be able to extend, too. To be able to find the error, we would have to see a reproducible example.

Anyway, this seems more like an issue of PHP skill rather than mPDF-specific question.

**Reply by finwe** [reply]
Given that you do not call WriteHTML in your examples, my wild guess about the empty page would be that you don't call AddPage first, which you have to manually in this case.

**Reply by sscherger** [reply]
If I don't have to, so I would not extend mPDF. My main thing is to have some functions and to pass all the parameters correctly. 

So far I did not use HTML, so I thought AddPage() would not be needed, but I saw that only WriteHTML adds a page automatically. When adding AddPage() I see this warning:
```
Warning: Undefined array key 1 in C:\Apache24\htdocs\vendor\mpdf\mpdf\src\Writer\MetadataWriter.php on line 559
Warning: Trying to access array offset on value of type null in C:\Apache24\htdocs\vendor\mpdf\mpdf\src\Writer\MetadataWriter.php on line 561
Warning: Trying to access array offset on value of type null in C:\Apache24\htdocs\vendor\mpdf\mpdf\src\Writer\MetadataWriter.php on line 567
Warning: Trying to access array offset on value of type null in C:\Apache24\htdocs\vendor\mpdf\mpdf\src\Writer\MetadataWriter.php on line 567
Error detected. PDF file generation aborted: Trying to access array offset on value of type null 
```

Also looks like the function can not be found. In FPDF and TCPDF I did  it like that:

```
<?php
require('tcpdf.php');
class print extends TCPDF{
	function Address($x,$y,$company){		
		$this->SetY($y); 	
		$this->SetX($x);
		$this->MultiCell(85,3, ($company),0,'L',0,1); ...
}
new print
$pdf->AddPage();
$pdf->address($x,$y,$company) 
...
$pdf->Output();
?> 
```

But, how do I do that when I encapsulate the object like you write: 
`function printAddress(Mpdf $mpdf, $x, $y, $company)`
I have never seen a line lke that before. How do I call that function? I got only errors with my tries. 

Sorry for that! We have been working here for 15 years with the programme that I wrote for our company, but since I have never really learned it, I still learn very basic things from time to time that I have never been confronted with before - and then unfortunately I find it very difficult at that moment. 
If I can create a PDF with a function here, I definitely have the tools to create the form with mPDF. I started this morning to continue working on the problem right away, but I would be very happy if you could please help me a little!

**Reply by finwe** [reply]
If you are experiencing warnings from mpdf, please create a new Issue with reproducible example HTML/PHP code. Or share the code causing this here. It is not possible to advise unless we see the code.

As for the [composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance): you call it just like you would any other function in your code:

```
use Mpdf\Mpdf;

function printAddress(Mpdf $mpdf, $x, $y, $company) 
{
    $mpdf->SetY($y); 	
    $mpdf->SetX($x);
    $mpdf->MultiCell(85,3, ($company),0,'L',0,1);
}

$mpdf = new Mpdf();
$mpdf->AddPage();

printAddress($mpdf, 0, 0, 'Company name');

$mpdf->Output();
```

Or you might encapsulate all this to a class and promote the mPDF object to a property:

```
class PdfGenerator
{

    private Mpdf $mpdf;

    public function __construct()
    {
        $this->mpdf = new Mpdf();
        $this->mpdf->AddPage();
    }

    public function printAddress($x, $y, $company) 
    {
        $this->mpdf->SetY($y); 	
        $this->mpdf->SetX($x);
        $this->mpdf->MultiCell(85,3, ($company),0,'L',0,1);
    }

    public function output()
    {
        return $this->mpdf->Output();
    }

}

$generator = new PdfGenerator();
$generator->printAddress(0, 0, 'Company name');
$generator->output();
```

**Reply by sscherger** [reply]
Wow! Thank you very much for that code!
I tried both. 

With the first I get the address on the page!:
```
<?php
	use \Mpdf\Mpdf;
try {
	require_once ('../../vendor/autoload.php');
	require_once('../konf/konf.ink.php');
	define('TCPDF_FONTPATH', 'font/');
	$db = new mysqli(DBHOST, DBBENUTZER, DBPASSWORT, DBNAME);

	function printAddress(Mpdf $mpdf, $x,$y,$RFirma1){
	
		$mpdf->SetY($y); 
		$mpdf->SetFont('Helvetica','',10);
		$mpdf->SetX($x);
		$mpdf->MultiCell(85,3, ($RFirma1),0,'L',false);
                ...
	}

$invoice= ((!empty ($_GET['invoice'])) ? $_GET['invoice'] : ''); 
$order= ((!empty ($_GET['order'])) ? $_GET['order'] : ''); 

$query = $db->query("SELECT  ..." );
if (!$query) {
	die("Execute query error, because: ". $db->error);
}
$result = $query->fetch_object();	

$mpdf = new Mpdf([
	'PDFA' => true,
	'PDFAauto' => true,
	'margin_left' => 20,
	'margin_right' => 15,
	'margin_top' => 48,
	'margin_bottom' => 25,
	'margin_header' => 10,
	'margin_footer' => 10
]);

$mpdf ->debug = true;

$mpdf->AddPage();

$mpdf->WriteCell(26,5,'Test:',0,0,'L',0);
printAddress($mpdf, 80,63,$result->RFirma1);

$mpdf ->Output();  
	
} catch (\Mpdf\MpdfException $e) { 
echo $e->getMessage();
}
?> 
```

With the second I get the address on the paper, but the "Test" will not be printed and an error appears:
```
<?php
	use \Mpdf\Mpdf;
try {
	require_once ('../../vendor/autoload.php');
	require_once('../konf/konf.ink.php');
	define('TCPDF_FONTPATH', 'font/');
	$db = new mysqli(DBHOST, DBBENUTZER, DBPASSWORT, DBNAME);

class PdfGenerator {

    private Mpdf $mpdf;

    public function __construct()
    {
        $this->mpdf = new Mpdf();
        $this->mpdf->AddPage();
    }
	
	public function printAddress($x,$y,$RFirma1,$RFirma2,$RFirma3,$RStrasse,$RPLZ,$RStadt,$RLand){		// Adresse
		$this->mpdf->SetY($y); 
		$this->mpdf->SetFont('Helvetica','',10);
		$this->mpdf->SetX($x);
		$this->mpdf->MultiCell(85,3, ($RFirma1),0,'L',false);
		if ((isset($RFirma2) || ($RFirma2 <> '')) && ($RFirma2 <> '-') && $RFirma2 != null){
			$this->mpdf->SetX($x);
			$this->mpdf->MultiCell(85,3, ($RFirma2),0,'L',false);
			if ((isset($RFirma3) || ($RFirma3 <> '')) && ($RFirma3 <> '-') && $RFirma3 != null){
				$this->mpdf->SetX($x);
				$this->mpdf->MultiCell(85,3, ($RFirma3),0,'L',false);
			}
		}
		$this->mpdf->SetX($x);
		$this->mpdf->MultiCell(85,3, ($RStrasse),0,'L',false);
		$this->mpdf->SetX($x);
		$this->mpdf->WriteCell(85,3, ($RPLZ).'  '. ($RStadt),0,0,'L',false);
		$this->mpdf->Ln();
		$this->mpdf->SetX($x);
		if( ($RLand) == 'Germany' ||  ($RLand) == 'GERMANY' ||  ($RLand) == '-' || ($RLand) == '---'){
		}else{
			$this->mpdf->WriteCell(85,3, ($RLand),0,0,'L',0);
		}
		$this->mpdf->SetFont('Helvetica','',8);
	}
    public function output()
    {
        return $this->mpdf->Output();
    }
}

$rechnungsnr = ((!empty ($_GET['rechnungsnr'])) ? $_GET['rechnungsnr'] : ''); 
$auftragsnr = ((!empty ($_GET['auftragsnr'])) ? $_GET['auftragsnr'] : ''); 

$query = $db->query("SELECT   .." );
if (!$query) {
	die("Execute query error, because: ". $db->error);
}
$result = $query->fetch_object();	

$generator  = new PdfGenerator ([
	'PDFA' => true,
	'PDFAauto' => true,
	'margin_left' => 20,
	'margin_right' => 15,
	'margin_top' => 48,
	'margin_bottom' => 25,
	'margin_header' => 10,
	'margin_footer' => 10
]);

$generator ->debug = true;

$generator->AddPage();

$generator->WriteCell(26,5,'Test:',0,0,'L',0);
$generator ->printAddress(80,63,$result->RFirma1);

$generator ->Output();  	
} catch (\Mpdf\MpdfException $e) { 
echo $e->getMessage();
}
?> 
```
The message: 
`Fatal error: Uncaught Error: Call to undefined method PdfGenerator::AddPage() in C:\Apache24\htdocs\ahock\mpdf\druckeRechnungDeu.php:130 Stack trace: #0 {main} thrown in C:\Apache24\htdocs\ahock\mpdf\druckeRechnungDeu.php on line 130`
Do I have to extend the PdfGenerator with Mpdf? 

Thank you very much for your great support! Think I'm back in track!

**Reply by finwe** [reply]
You do not call the AddPage on the generator object, you need to call addpage inside the class on the `$this->mpdf` property (or you don't need to call it at all, because it is already called in the constructor - see below). 

Furthermore, when you pass the parameters to the PdfGenerator, you need to pass the parameter to the mpdf object inside. 

    public function __construct(array $config)
    {
        $this->mpdf = new Mpdf($config);
        $this->mpdf->AddPage();
    }
